File: cpp/samchon/examples/packer/WrapperArray.cpp
#include <samchon/examples/packer/WrapperArray.hpp>

#include <samchon/library/FactorialGenerator.hpp>

using namespace std;
using namespace samchon::library;
using namespace samchon::example::packer;

/* ---------------------------------------------------------
	CONSTRUCTOR
--------------------------------------------------------- */
WrapperArray::WrapperArray()
	: super()
{
	this->reserved = make_shared<ProductArray>();
	this->sample = make_shared<Wrapper>();
}
WrapperArray::WrapperArray(const string &name, int price, int volume, int weight)
	: super()
{
	this->reserved = make_shared<ProductArray>();
	this->sample = make_shared<Wrapper>(name, price, volume, weight);
}
WrapperArray::WrapperArray(const WrapperArray &wrapperArray)
	: super()
{
	this->reserved = make_shared<ProductArray>();
	this->sample = wrapperArray.sample;
}
void WrapperArray::construct(shared_ptr<XML> xml)
{
	super::construct(xml);

	sample->construct(xml);
}

auto WrapperArray::createChild(shared_ptr<XML>) -> Wrapper*
{
	return new Wrapper();
}

/* ---------------------------------------------------------
	CALCULATE AND OPTIMIZE
--------------------------------------------------------- */
auto WrapperArray::tryInsert(shared_ptr<Product> product) -> bool
{
	if (product->getVolume() > sample->getVolume() ||
		product->getWeight() > sample->getWeight())
	{
		return false;
	}
	
	reserved->push_back(product);
	return true;
}
void WrapperArray::optimize()
{
	if(reserved->empty() == true)
		return;
	
	FactorialGenerator factorial(reserved->size());
	shared_ptr<WrapperArray> minWrapperArray = nullptr;
	
	for (int i = 0; i < factorial.size(); i++)
	{
		shared_ptr<WrapperArray> wrapperArray(new WrapperArray(*this));
		vector<size_t> &row = factorial[i];

		for (size_t j = 0; j < row.size(); j++)
		{
			shared_ptr<Product> &product = this->reserved->at(row[j]);
			
			if (wrapperArray->empty() == true || 
				wrapperArray->at(wrapperArray->size() - 1)->tryInsert(product) == false)
			{
				Wrapper *wrapper = new Wrapper(*this->sample);
				wrapper->tryInsert(product);

				wrapperArray->emplace_back(wrapper);
			}
		}
		
		//unique_lock<mutex> uk(mtx);
		if (minWrapperArray == nullptr ||
			wrapperArray->size() < minWrapperArray->size())
		{
			minWrapperArray = wrapperArray;
		}
	}

	assign(minWrapperArray->begin(), minWrapperArray->end());
}
auto WrapperArray::calcPrice() const -> int
{
	return sample->getPrice() * (int)size();
}

/* ---------------------------------------------------------
	EXPORT
--------------------------------------------------------- */
auto WrapperArray::TAG() const -> string
{
	return "wrapperArray";
}
auto WrapperArray::CHILD_TAG() const -> string
{
	return "wrapper";
}

auto WrapperArray::toXML() const -> shared_ptr<XML>
{
	shared_ptr<XML> &xml = super::toXML();
	xml->addAllProperty(sample->toXML());

	return xml;
}

auto WrapperArray::toString() const -> string
{
	string str = "Category - " + sample->getName() + "\n";
	for (size_t i = 0; i < size(); i++)
		str += at(i)->toString() + "\n";

	return move(str);
}

File: ts/samchon/example/packer/WrapperArray.ts
/// <reference path="../../API.ts" />

/// <reference path="../../protocol/EntityArray.ts" />
///     <reference path="Wrapper.ts" />

namespace samchon.example.packer
{
    export class WrapperArray
        extends protocol.EntityArray<Wrapper>
    {
        /**
         * <p> A list for reserved Product(s). </p>
         */
        private reserved: Array<Product>;

        /**
         * <p> A sample wrapper used to copy. </p>
         */
        private sample: Wrapper;
    
        /* --------------------------------------------------------------------
            CONSTRUCTORS
        -------------------------------------------------------------------- */
        /**
	     * <p> Construct from a sample wrapper. </p>
	     *
	     * @param sample A sample wrapper used to copy wrappers.
	     */
        public constructor(sample: Wrapper = null)
        {
            super();

            this.sample = sample;
            this.reserved = new Array<Product>();
        }
        public construct(xml: library.XML): void
        {
            super.construct(xml);

            this.sample = new Wrapper();
            this.sample.construct(xml);
        }

        protected createChild(xml: library.XML): Wrapper
        {
            return new Wrapper();
        }

        /* --------------------------------------------------------------------
            OPERATORS
        -------------------------------------------------------------------- */
        /**
	     * <p> Try to insert a product into reserved list. </p>
	     *
	     * <p> If the Product's volume and weight is equal or less than the Wrapper categorized so that enable to
	     * insert in a Wrapper, reserve the Product and returns <i>true</i>. If not, does not reserve and just
	     * return <i>false</i>. </p>
	     *
	     * @return Whether the Product's volume and weight is equal or less than the Wrapper.
	     */
        public tryInsert(product: Product): boolean
        {
            if (product.getVolume() > this.sample.getVolume() ||
                product.getWeight() > this.sample.getWeight())
            {
                return false;
            }

            this.reserved.push(product);
            return true;
        }

        /**
	     * <p> Optimize to retrieve the best solution. </p>
	     *
	     * <p> Retrieves the best solution of packaging in level of WrapperArray. </p>
	     * <p> Shuffles sequence of reserved Product(s) by samchon::library::FactorialGenerator and insert the reserved
	     * Products(s) following the sequence creating Wrapper(s) as needed. Between the sequences from FactorialGenerator,
	     * retrieve and determine the best solution. </p>
	     *
	     * <h4> Note. </h4>
	     * <p> Sequence of inserting Product can affeact to numbers of Wrapper(s) to be used. </p>
	     * <p> It's the reason why even WrapperArray has the optimize() method. </p>
	     */
        public optimize(): void
        {
            if (this.reserved.length == 0)
                return;

            var factorial: library.FactorialGenerator = new library.FactorialGenerator(this.reserved.length);
            var minWrapperArray: WrapperArray;

            for (var i: number = 0; i < factorial.size(); i++)
            {
                var wrapperArray: WrapperArray = new WrapperArray(this.sample);
                var row: Array<number> = factorial.at(i);

                for (var j: number = 0; j < row.length; j++)
                {
                    var product: Product = this.reserved[row[j]];

                    if (wrapperArray.size() == 0 ||
                        wrapperArray.at(wrapperArray.size() - 1).tryInsert(product) == false)
                    {
                        var wrapper: Wrapper = new Wrapper(this.sample);
                        wrapper.tryInsert(product);

                        wrapperArray.push(wrapper);
                    }
                }

                if (minWrapperArray == null ||
                    wrapperArray.calcPrice() < minWrapperArray.calcPrice())
                {
                    minWrapperArray = wrapperArray;
                }
            }

            //REPLACE TO MIN_WRAPPER_ARRAY
            this.splice(0, this.size());

            for (var i: number = 0; i < minWrapperArray.size(); i++)
                this.push( minWrapperArray.at(i) );
        }

        /* --------------------------------------------------------------------
            GETTERS
        -------------------------------------------------------------------- */
        /**
         * <p> Calculate price of the Wrapper(s). </p>
         * 
         * <p> Calculates price of all wrappers'. The price does not contain inserted products'. </p>
         */
        public calcPrice(): number
        {
            return this.sample.getPrice() * this.size();
        }

        /**
         * <p> Get sample. </p>
         */
        public getSample(): Wrapper
        {
            return this.sample;
        }

        /* --------------------------------------------------------------------
            EXPORTERS
        -------------------------------------------------------------------- */
        public TAG(): string 
        {
            return "wrapperArray";
        }
        public CHILD_TAG(): string 
        {
            return "wrapper";
        }

        public toXML(): library.XML
        {
            var xml: library.XML = super.toXML();
            xml.addAllProperties(this.sample.toXML());

            return xml;
        }
    }
}